packages/dev/core/src/Materials/clipPlaneMaterialHelper.ts

```diff
@@ -0,0 +1,83 @@
+import type { Effect } from "./effect";
+import type { IClipPlanesHolder } from "../Misc/interfaces/iClipPlanesHolder";
+import type { Nullable } from "../types";
+import type { Plane } from "../Maths/math.plane";
+
+/** @internal */
+export function addClipPlaneUniforms(uniforms: string[]): void {
+    if (uniforms.indexOf("vClipPlane") === -1) {
+        uniforms.push("vClipPlane");
+    }
+    if (uniforms.indexOf("vClipPlane2") === -1) {
+        uniforms.push("vClipPlane2");
+    }
+    if (uniforms.indexOf("vClipPlane3") === -1) {
+        uniforms.push("vClipPlane3");
+    }
+    if (uniforms.indexOf("vClipPlane4") === -1) {
+        uniforms.push("vClipPlane4");
+    }
+    if (uniforms.indexOf("vClipPlane5") === -1) {
+        uniforms.push("vClipPlane5");
+    }
+    if (uniforms.indexOf("vClipPlane6") === -1) {
+        uniforms.push("vClipPlane6");
+    }
+}
+
+/** @internal */
+export function prepareDefinesForClipPlanes(primaryHolder: IClipPlanesHolder, secondaryHolder: IClipPlanesHolder, defines: string[] | Record<string, any>): boolean {
+    let changed = false;
+    let clipPlane = primaryHolder.clipPlane ?? secondaryHolder.clipPlane;
+    addDefine(clipPlane, defines, "CLIPPLANE", changed);
+    clipPlane = primaryHolder.clipPlane2 ?? secondaryHolder.clipPlane2;
+    changed = addDefine(clipPlane, defines, "CLIPPLANE2", changed);

```

Comment:
nit: I would do `changed ||= addDefine(...)` and not pass `changed` to `addDefine` as this parameter is not really used in the function.

Reply:
this would prevent the call to addDefine if changed is already true https://jsfiddle.net/yv8rLhbn/

Reply:
Indeed, so that would be `changed = addDefine(...) || changed` instead.